// llparse/lib/implementation/c/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SIGNED_TYPES = exports.UNSIGNED_TYPES = exports.UNSIGNED_LIMITS = exports.SIGNED_LIMITS = exports.SEQUENCE_PAUSE = exports.SEQUENCE_MISMATCH = exports.SEQUENCE_COMPLETE = exports.VAR_MATCH = exports.ARG_ENDPOS = exports.ARG_POS = exports.ARG_STATE = exports.BLOB_PREFIX = exports.STATE_ERROR = exports.STATE_PREFIX = exports.LABEL_PREFIX = exports.CONTAINER_KEY = void 0;
exports.CONTAINER_KEY = 'c';
exports.LABEL_PREFIX = '';
exports.STATE_PREFIX = 's_n_';
exports.STATE_ERROR = 's_error';
exports.BLOB_PREFIX = 'llparse_blob';
exports.ARG_STATE = 'state';
exports.ARG_POS = 'p';
exports.ARG_ENDPOS = 'endp';
exports.VAR_MATCH = 'match';
// MatchSequence
exports.SEQUENCE_COMPLETE = 'kMatchComplete';
exports.SEQUENCE_MISMATCH = 'kMatchMismatch';
exports.SEQUENCE_PAUSE = 'kMatchPause';
exports.SIGNED_LIMITS = new Map();
exports.SIGNED_LIMITS.set('i8', ['-0x80', '0x7f']);
exports.SIGNED_LIMITS.set('i16', ['-0x8000', '0x7fff']);
exports.SIGNED_LIMITS.set('i32', ['(-0x7fffffff - 1)', '0x7fffffff']);
exports.SIGNED_LIMITS.set('i64', ['(-0x7fffffffffffffffLL - 1)',
    '0x7fffffffffffffffLL']);
exports.UNSIGNED_LIMITS = new Map();
exports.UNSIGNED_LIMITS.set('i8', ['0', '0xff']);
exports.UNSIGNED_LIMITS.set('i8', ['0', '0xff']);
exports.UNSIGNED_LIMITS.set('i16', ['0', '0xffff']);
exports.UNSIGNED_LIMITS.set('i32', ['0', '0xffffffff']);
exports.UNSIGNED_LIMITS.set('i64', ['0ULL', '0xffffffffffffffffULL']);
exports.UNSIGNED_TYPES = new Map();
exports.UNSIGNED_TYPES.set('i8', 'uint8_t');
exports.UNSIGNED_TYPES.set('i16', 'uint16_t');
exports.UNSIGNED_TYPES.set('i32', 'uint32_t');
exports.UNSIGNED_TYPES.set('i64', 'uint64_t');
exports.SIGNED_TYPES = new Map();
exports.SIGNED_TYPES.set('i8', 'int8_t');
exports.SIGNED_TYPES.set('i16', 'int16_t');
exports.SIGNED_TYPES.set('i32', 'int32_t');
exports.SIGNED_TYPES.set('i64', 'int64_t');
