// llparse/lib/implementation/c/code/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
const external_1 = require("./external");
const is_equal_1 = require("./is-equal");
const load_1 = require("./load");
const mul_add_1 = require("./mul-add");
const or_1 = require("./or");
const store_1 = require("./store");
const test_1 = require("./test");
const update_1 = require("./update");
__exportStar(require("./base"), exports);
exports.default = {
    IsEqual: is_equal_1.IsEqual,
    Load: load_1.Load,
    Match: class Match extends external_1.External {
    },
    MulAdd: mul_add_1.MulAdd,
    Or: or_1.Or,
    Span: class Span extends external_1.External {
    },
    Store: store_1.Store,
    Test: test_1.Test,
    Update: update_1.Update,
    Value: class Value extends external_1.External {
    },
};
