// bitcode-builder/lib/values/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const values = require("./");
class Value {
    constructor(ty) {
        this.ty = ty;
    }
    // Node.js REPL
    inspect() {
        return this.toString();
    }
    isArgument() {
        return this instanceof values.Argument;
    }
    isGlobal() {
        return this instanceof values.Global;
    }
    isBasicBlock() {
        return this instanceof values.BasicBlock;
    }
    isConstant() {
        return this instanceof values.constants.Constant;
    }
    isInstruction() {
        return this instanceof values.instructions.Instruction;
    }
    toArgument() {
        assert(this.isArgument(), 'Value is not an Argument instance');
        return this;
    }
    toGlobal() {
        assert(this.isGlobal(), 'Value is not an Global instance');
        return this;
    }
    toBasicBlock() {
        assert(this.isBasicBlock(), 'Value is not an BasicBlock instance');
        return this;
    }
    toConstant() {
        assert(this.isConstant(), 'Value is not an Constant instance');
        return this;
    }
    toInstruction() {
        assert(this.isInstruction(), 'Value is not an Instruction instance');
        return this;
    }
}
exports.Value = Value;
