// bitcode-builder/lib/values/instructions/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = require("../base");
const instructions = require("./");
class Instruction extends base_1.Value {
    constructor(ty, opcode, operands) {
        super(ty);
        this.opcode = opcode;
        this.operands = operands;
        this.metadata = new Map();
    }
    isExtractValue() {
        return this instanceof instructions.ExtractValue;
    }
    isInsertValue() {
        return this instanceof instructions.InsertValue;
    }
    isGetElementPtr() {
        return this instanceof instructions.GetElementPtr;
    }
    isUnreachable() {
        return this instanceof instructions.Unreachable;
    }
    *[Symbol.iterator]() {
        yield* this.operands;
    }
    toString() {
        const operands = this.operands.map((op) => {
            if (op instanceof Instruction) {
                return `[instr opcode=${op.opcode}]`;
            }
            return op.toString();
        });
        return `[instr opcode=${this.opcode} operands={${operands.join(', ')}}]`;
    }
}
exports.Instruction = Instruction;
