// bitcode-builder/lib/values/instructions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = require("./base");
exports.Instruction = base_1.Instruction;
var binop_1 = require("./binop");
exports.Binop = binop_1.Binop;
var branch_1 = require("./branch");
exports.Branch = branch_1.Branch;
var call_1 = require("./call");
exports.Call = call_1.Call;
var cast_1 = require("./cast");
exports.Cast = cast_1.Cast;
var extractvalue_1 = require("./extractvalue");
exports.ExtractValue = extractvalue_1.ExtractValue;
var getelementptr_1 = require("./getelementptr");
exports.GetElementPtr = getelementptr_1.GetElementPtr;
var icmp_1 = require("./icmp");
exports.ICmp = icmp_1.ICmp;
var insertvalue_1 = require("./insertvalue");
exports.InsertValue = insertvalue_1.InsertValue;
var jump_1 = require("./jump");
exports.Jump = jump_1.Jump;
var load_1 = require("./load");
exports.Load = load_1.Load;
var phi_1 = require("./phi");
exports.Phi = phi_1.Phi;
var ret_1 = require("./ret");
exports.Ret = ret_1.Ret;
var store_1 = require("./store");
exports.Store = store_1.Store;
var switch_1 = require("./switch");
exports.Switch = switch_1.Switch;
var unreachable_1 = require("./unreachable");
exports.Unreachable = unreachable_1.Unreachable;
