// llparse/lib/implementation/js/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNSIGNED_LIMITS = exports.SIGNED_LIMITS = exports.SEQUENCE_PAUSE = exports.SEQUENCE_MISMATCH = exports.SEQUENCE_COMPLETE = exports.VAR_MATCH = exports.ARG_OFF = exports.ARG_BUF = exports.ARG_CURRENT = exports.TABLE_PREFIX = exports.BLOB_PREFIX = exports.STATE_ERROR = exports.STATE_PREFIX = exports.CONTAINER_KEY = void 0;
exports.CONTAINER_KEY = 'js';
exports.STATE_PREFIX = 'S_N_';
exports.STATE_ERROR = 'S_ERROR';
exports.BLOB_PREFIX = 'BLOB_';
exports.TABLE_PREFIX = 'LOOKUP_TABLE_';
exports.ARG_CURRENT = 'current';
exports.ARG_BUF = 'buf';
exports.ARG_OFF = 'off';
exports.VAR_MATCH = 'match';
// MatchSequence
exports.SEQUENCE_COMPLETE = 'SEQUENCE_COMPLETE';
exports.SEQUENCE_MISMATCH = 'SEQUENCE_MISMATCH';
exports.SEQUENCE_PAUSE = 'SEQUENCE_PAUSE';
exports.SIGNED_LIMITS = new Map();
exports.SIGNED_LIMITS.set('i8', ['-0x80', '0x7f']);
exports.SIGNED_LIMITS.set('i16', ['-0x8000', '0x7fff']);
exports.SIGNED_LIMITS.set('i32', ['-0x80000000', '0x7fffffff']);
exports.SIGNED_LIMITS.set('i64', ['-0x8000000000000000n', '0x7fffffffffffffffn']);
exports.UNSIGNED_LIMITS = new Map();
exports.UNSIGNED_LIMITS.set('i8', ['0', '0xff']);
exports.UNSIGNED_LIMITS.set('i16', ['0', '0xffff']);
exports.UNSIGNED_LIMITS.set('i32', ['0', '0xffffffff']);
exports.UNSIGNED_LIMITS.set('i64', ['0n', '0xffffffffffffffffn']);
